### PR TITLE
Selfie

### LINKED ISSUE #969: Front Camera by default
I would like to request to add a feature to ODK for opening the front camera by default. In some cases the users may need to take their own snap as a record and the level of users are: they can hardly can mange to use couple of functions on a form and flipping camera could be big issue for them.

### PR DESCRIPTION
**Pull request by grzesiek2010**
This PR is in reference to resolve #969 

@lognaturel 
I've implemented changes to capture selfie.

> they can hardly can mange to use couple of functions on a form and flipping camera could be big issue for them

Following above advise I've decided to do that as easy as possible (as it's for users who are not very much comfortable with technology) so there is no additional buttons when user open the camera we just need to tap the screen to take a selfie. 

Let me know what you think. If you like this approach I would like to test my changes on various devices to be sure everything works well.

Here is a form for testing: 
[images.txt](https://github.com/opendatakit/collect/files/992870/images.txt)

### CODE REVIEW

**Comment by lognaturel** [comment]
@grzesiek2010 Did you say that you got this working with a Galaxy 4 on Android 5.0.1? Unfortunately it doesn't seem to work either on my Galaxy 4 running Android 7 or my Nexus 5 running Android 6.0.1 so it's hard to evaluate. More broadly, we do need to carefully consider whether we want to add things to the core that we know only work on some devices.

@imran3may what devices did you say you would be using?

**Comment by grzesiek2010** [comment]
@lognaturel I was talking about small hack:

> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
intent.putExtra("android.intent.extras.LENS_FACING_FRONT", 1);
} else {
intent.putExtra("android.intent.extras.CAMERA_FACING", 1);
}

but as I said it doesn't work and I didn't use it.
Instead of that I implemented this solution and it should work on all devices. I've tested it using at least three devices and everything was ok. You said you use Android 7 so I guess it may be the reason as some used code is deprecated. I'll try to add new code for newer versions of android.

**Comment by grzesiek2010** [comment]
@lognaturel I've just tested Sony Z3 Android 6 and it works as well :) Are you sure you use an appropriate widget? There are two widgets in that test form - first normal image capture and the second for selfie.

**Comment by grzesiek2010** [comment]
@lognaturel 
I've added new code for android version >= 21 so please check it now and let me know.

**Comment by lognaturel** [comment]
😭😭 You are exactly right, I didn't swipe far enough to even get to the selfie mode. I was looking at the code going "this really should work!" 😖

Good news is that it works on both devices with the update, yay!

I think the button for taking a picture can use the same "Take Picture" text as the normal picture taking one. "Selfie" is hard to translate in different languages and may not be understood. Same for the toast, I think "Tap the screen to take a picture" will be easier to understand. Also, since the idea is to force the user to take an up-to-date photo, the "Choose Image" button should probably be hidden.

**Comment by grzesiek2010** [comment]
@lognaturel ok I'll add those minor changes and I'll test it using various devices again.

**Comment by lognaturel** [comment]
Thanks! Another thing -- for both of my devices, although the picture is always taken in portrait, it is always shown below the buttons in landscape. I know dealing with screen rotation is a pain but it would be nice to see whether it's possible to support both portrait and landscape.

**Comment by lognaturel** [comment]
We just switched out the member name convention from under you 😳 so before you start working on things I'd recommend merging from master!

I also just got 
```
java.lang.RuntimeException: takePicture failed
	at android.hardware.Camera.native_takePicture(Native Method)
	at android.hardware.Camera.takePicture(Camera.java:1434)
	at android.hardware.Camera.takePicture(Camera.java:1379)
	at org.odk.collect.android.activities.CaptureSelfieActivity$1.onClick(CaptureSelfieActivity.java:64)
	at android.view.View.performClick(View.java:5204)
	at android.view.View$PerformClick.run(View.java:21153)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by imran3may** [comment]
Dear Hélène,

We are going to use Samsung Galaxy duos for this purpose.

Thanks,
Imran


A N M AL IMRAN (Imran)
Brossard, QC
Canada
Cell: +1 (438) 501 3576

**Comment by grzesiek2010** [comment]
@lognaturel 

> I think the button for taking a picture can use the same "Take Picture" text as the normal picture taking one. "Selfie" is hard to translate in different languages and may not be understood. Same for the toast, I think "Tap the screen to take a picture" will be easier to understand. Also, since the idea is to force the user to take an up-to-date photo, the "Choose Image" button should probably be hidden.

Fixed.

> Another thing -- for both of my devices, although the picture is always taken in portrait, it is always shown below the buttons in landscape. I know dealing with screen rotation is a pain but it would be nice to see whether it's possible to support both portrait and landscape.

Please try to take a photo (not selfie) then you notice the same behavior. I agree we should improve that but not as a part of this issue.

Could you try to reproduce the crash you mentioned now. I hope everything is ok.

**Comment by lognaturel** [comment]
Looking good!

I should have mentioned that I got that crash running `adb shell monkey -p org.odk.collect.android -v 5000`. I'm not able to reproduce it now so I think we're probably fine. I think the simplified interface with just the take picture button is nice.

On both of my devices, the orientation is actually properly maintained for previews for the normal image widget. That is, if I take a picture in portrait, the preview is in portrait and if I take it in landscape, the preview is in landscape. Is that not the case for you?

Did you get a chance to look into supporting both portrait and landscape?

**Comment by grzesiek2010** [comment]
@lognaturel 

> On both of my devices, the orientation is actually properly maintained for previews for the normal image widget. That is, if I take a picture in portrait, the preview is in portrait and if I take it in landscape, the preview is in landscape. Is that not the case for you?

I've just tested my devices and
It works as I described if I use: 

Samsung Galaxy S4 Android 5.0.1
Samsung Duos Android 4.2.2
Sony Xperia Z3 ANdroid 6.0.1
Samsung Galaxy J1 Android 4.4.4

everythink works ok (as you described) only when I use:
Huawei Y560-LO1 Android 5.0.1

so as I said it's a more common problem and we should create a separate issue for it.

> Did you get a chance to look into supporting both portrait and landscape?

not yet I'll start today.

**Comment by grzesiek2010** [comment]
@lognaturel 

> Did you get a chance to look into supporting both portrait and landscape?

Now it supports both orientations. I just need to test all changes using various devices and everything should be ok I think.

**Comment by lognaturel** [comment]
It rotates, yay!

On the Moto G4, Android 7.0.0, portrait mode works perfectly. Landscape works but the preview is upside down!

On the Nexus 5, Android 6.0.1, I get the following crash taking pictures in both orientations:
```
FATAL EXCEPTION: CameraBackground
Process: org.odk.collect.android, PID: 7116
java.lang.IllegalStateException: maxImages (2) has already been acquired, call #close before acquiring more.
	at android.media.ImageReader.acquireNextImage(ImageReader.java:383)
	at org.odk.collect.android.fragments.Camera2Fragment$3.onImageAvailable(Camera2Fragment.java:241)
	at android.media.ImageReader$ListenerHandler.handleMessage(ImageReader.java:648)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:148)
	at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by grzesiek2010** [comment]
@lognaturel 

> On the Nexus 5, Android 6.0.1, I get the following crash taking pictures in both orientations

Using my intuition (and stackoverflow :)) I've tried to fix this bug so please check it.

Today I don't have a lot of devices so I've tested all changes using:
Samsung Galaxy S4 Android 5.0.1
Infinix X451 Android 4.2.1
Samsung Duos Android 4.2.2
Everything is ok using above mentioned devices.

Hopefully tomorrow I will have more devices so the I finish testing and try to fix :

> On the Moto G4, Android 7.0.0, portrait mode works perfectly. Landscape works but the preview is upside down!

**Comment by lognaturel** [comment]
😪 Stackoverflow let us down this time! I'm still getting the same crash. I can try to look into it a bit tomorrow but I don't think this is a deal breaker. The Nexus 5 seems to have some special and strange behavior.

File: collect_app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -524,4 +524,7 @@
 <string name="generate">Generate</string>
 <string name="invalid_qrcode">QR code does not contain valid settings</string>
     <string name="share_qrcode">Share QR code</string>
+    <string name="take_picture_instruction">Tap the screen to take a picture</string>
+    <string name="request_permission">This application needs camera permission.</string>
+    <string name="camera_error">This device doesn\'t support Camera2 API.</string>
```
Comment (lognaturel) [inline collect_app/src/main/res/values/strings.xml]: support *the* Camera2 API

**Comment by grzesiek2010** [comment]
@lognaturel 
I've been testing using various devices but I'm not able to reproduce this issue:

> On the Moto G4, Android 7.0.0, portrait mode works perfectly. Landscape works but the preview is upside down!

I think it's a problem with that particular device eg Nexus had the same problem https://www.theverge.com/2015/11/9/9696774/google-nexus-5x-upside-down-camera
But now it's ok whe you user your Nexus right?

As I mentioned I can reproduce the issue only if I rotate my device form landscape to landscape (180 degrees) really quickly (and everything is ok if I rotate to portrait mode again) - it's not a common case I believe so we can try to fix that in the future.

**Comment by lognaturel** [comment]
I can prevent a hard crash on the Nexus 5 by catching the exception so I added that in. It will also give us a chance to see how common it is. 

The preview is also upside down when the picture is taken in landscape on the Nexus 5 but we can file that as a follow on issue.

There's a pretty long lag between taking the picture and seeing the preview. We should look into speeding it up and/or adding a visual queue that the image is saving. I imagine it's really long on a slower device.

@yanokwa since I added to this branch, please do a final review!

**Review by lognaturel — APPROVED** [review]

File: collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    private Camera getCameraInstance() {
+        Camera camera = null;
+        for (int camNo = 0; camNo < Camera.getNumberOfCameras(); camNo++) {
+            Camera.CameraInfo camInfo = new Camera.CameraInfo();
+            Camera.getCameraInfo(camNo, camInfo);
+
+            if (camInfo.facing == (Camera.CameraInfo.CAMERA_FACING_FRONT)) {
```
Comment (yanokwa) [inline collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java:99]: @grzesiek2010 This isn't stopping a merge, but why is this wrapped by ()

Comment (dcbriccetti) [reply collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java:99]: The curves of the parentheses are meant to put one in the mindset of a round camera lens. Comes from recent subliminal self-documentation research.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java:99]: @yanokwa It's not necessary you are right I missed it.

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java:99]: @dcbriccetti 😅

@grzesiek2010 Please issue a pull request to clean this up when you get a chance. It feels weird to issue one for something this small but I think we should get in the habit of cleaning/refactoring code as we go, even if it's small things!

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java:99]: @lognaturel  https://github.com/opendatakit/collect/pull/1055 done

**Pull request merged**